File: lib/utils/password_rules.dart
/// Shared cloud-password rules for registration and reset (must match Laravel).
class PasswordRules {
  PasswordRules._();

  static const int minLength = 8;
  static final RegExp _letter = RegExp(r'[A-Za-z]');
  static final RegExp _digit = RegExp(r'[0-9]');
  static final RegExp _symbol = RegExp(r'[^A-Za-z0-9]');

  static const String requirementHint =
      'At least 8 characters, with a letter, a number, and a symbol (e.g. !@#\$%)';

  /// Null when valid; otherwise a teacher-facing error.
  static String? validationError(String password) {
    final value = password;
    if (value.length < minLength) {
      return 'Password must be at least $minLength characters.';
    }
    if (!_letter.hasMatch(value)) {
      return 'Password must include at least one letter.';
    }
    if (!_digit.hasMatch(value)) {
      return 'Password must include at least one number.';
    }
    if (!_symbol.hasMatch(value)) {
      return 'Password must include at least one special character (e.g. !@#\$%).';
    }
    return null;
  }

  static bool isValid(String password) => validationError(password) == null;
}

File: omr_web/src/lib/auth/password-rules.ts
/** Cloud password rules — must match Laravel Password::defaults(). */
export const PASSWORD_MIN_LENGTH = 8;

export const PASSWORD_REQUIREMENT_HINT =
  "At least 8 characters, with a letter, a number, and a symbol (e.g. !@#$)";

const letterRe = /[A-Za-z]/;
const digitRe = /[0-9]/;
const symbolRe = /[^A-Za-z0-9]/;

/** Null when valid; otherwise a teacher-facing error. */
export function passwordValidationError(password: string): string | null {
  if (password.length < PASSWORD_MIN_LENGTH) {
    return `Password must be at least ${PASSWORD_MIN_LENGTH} characters.`;
  }
  if (!letterRe.test(password)) {
    return "Password must include at least one letter.";
  }
  if (!digitRe.test(password)) {
    return "Password must include at least one number.";
  }
  if (!symbolRe.test(password)) {
    return "Password must include at least one special character (e.g. !@#$).";
  }
  return null;
}

export function isValidPassword(password: string): boolean {
  return passwordValidationError(password) === null;
}
